File: frontend/src/store/guidedBuild.ts
import { create } from 'zustand';

export type GuidedBuildStep =
  | 'strategy'
  | 'colors'
  | 'core'
  | 'support'
  | 'mana_base'
  | 'sideboard'
  | 'review';

export interface ArchetypeOption {
  name: string;
  description: string;
  playstyle: string;
  meta_percentage?: number;
  example_cards: string[];
}

export interface ColorOption {
  colors: string[];
  name: string;
  description: string;
  strengths: string[];
  weaknesses: string[];
}

export interface CardRecommendation {
  card_name: string;
  card_id?: string;
  quantity: number;
  role: string;
  reasoning: string;
  image_uri?: string;
  mana_cost?: string;
  type_line?: string;
}

export interface CardSlotGroup {
  slot_name: string;
  description: string;
  target_count: number;
  recommendations: CardRecommendation[];
}

export interface LandRecommendation {
  card_name: string;
  card_id?: string;
  quantity: number;
  category: string;
  reasoning: string;
  image_uri?: string;
}

export interface SideboardRecommendation {
  card_name: string;
  card_id?: string;
  quantity: number;
  target_matchups: string[];
  reasoning: string;
  image_uri?: string;
}

export interface StepResponse {
  session_id: string;
  current_step: GuidedBuildStep;
  step_index: number;
  total_steps: number;
  step_title: string;
  step_description: string;
  data: Record<string, any>;
  ai_message: string;
}

export interface CompleteResponse {
  session_id: string;
  deck_id?: string;
  deck_name: string;
  main_deck: Record<string, any>[];
  sideboard: Record<string, any>[];
  strategy_summary: string;
  archetype: string;
  colors: string[];
  format: string;
  is_valid: boolean;
  validation_errors: string[];
}

interface GuidedBuildState {
  sessionId: string | null;
  currentStep: GuidedBuildStep | null;
  stepIndex: number;
  totalSteps: number;
  stepTitle: string;
  stepDescription: string;
  stepData: Record<string, any>;
  aiMessage: string;
  isLoading: boolean;
  error: string | null;

  // User selections accumulated across steps
  selectedArchetype: string | null;
  selectedColors: string[];
  selectedCards: Record<string, any>[];
  selectedLands: Record<string, any>[];
  selectedSideboard: Record<string, any>[];
  format: string;

  // Completed deck
  completedDeck: CompleteResponse | null;

  // Actions
  setStepResponse: (response: StepResponse) => void;
  setSelectedArchetype: (archetype: string) => void;
  setSelectedColors: (colors: string[]) => void;
  setSelectedCards: (cards: Record<string, any>[]) => void;
  setSelectedLands: (lands: Record<string, any>[]) => void;
  setSelectedSideboard: (cards: Record<string, any>[]) => void;
  setFormat: (format: string) => void;
  setLoading: (loading: boolean) => void;
  setError: (error: string | null) => void;
  setCompletedDeck: (deck: CompleteResponse | null) => void;
  reset: () => void;
}

const initialState = {
  sessionId: null,
  currentStep: null as GuidedBuildStep | null,
  stepIndex: 0,
  totalSteps: 7,
  stepTitle: '',
  stepDescription: '',
  stepData: {},
  aiMessage: '',
  isLoading: false,
  error: null as string | null,
  selectedArchetype: null as string | null,
  selectedColors: [] as string[],
  selectedCards: [] as Record<string, any>[],
  selectedLands: [] as Record<string, any>[],
  selectedSideboard: [] as Record<string, any>[],
  format: 'standard',
  completedDeck: null as CompleteResponse | null,
};

export const useGuidedBuildStore = create<GuidedBuildState>((set) => ({
  ...initialState,

  setStepResponse: (response) =>
    set({
      sessionId: response.session_id,
      currentStep: response.current_step,
      stepIndex: response.step_index,
      totalSteps: response.total_steps,
      stepTitle: response.step_title,
      stepDescription: response.step_description,
      stepData: response.data,
      aiMessage: response.ai_message,
      error: null,
    }),

  setSelectedArchetype: (archetype) => set({ selectedArchetype: archetype }),
  setSelectedColors: (colors) => set({ selectedColors: colors }),
  setSelectedCards: (cards) => set({ selectedCards: cards }),
  setSelectedLands: (lands) => set({ selectedLands: lands }),
  setSelectedSideboard: (cards) => set({ selectedSideboard: cards }),
  setFormat: (format) => set({ format }),
  setLoading: (isLoading) => set({ isLoading }),
  setError: (error) => set({ error }),
  setCompletedDeck: (completedDeck) => set({ completedDeck }),
  reset: () => set(initialState),
}));

File: backend/app/schemas/guided_build.py
from typing import Optional, List, Dict, Any
from uuid import UUID
from enum import Enum

from pydantic import BaseModel, Field


class GuidedBuildStep(str, Enum):
    STRATEGY = "strategy"
    COLORS = "colors"
    CORE = "core"
    SUPPORT = "support"
    MANA_BASE = "mana_base"
    SIDEBOARD = "sideboard"
    REVIEW = "review"


STEP_ORDER = [
    GuidedBuildStep.STRATEGY,
    GuidedBuildStep.COLORS,
    GuidedBuildStep.CORE,
    GuidedBuildStep.SUPPORT,
    GuidedBuildStep.MANA_BASE,
    GuidedBuildStep.SIDEBOARD,
    GuidedBuildStep.REVIEW,
]


class ArchetypeOption(BaseModel):
    name: str
    description: str
    playstyle: str
    meta_percentage: Optional[float] = None
    example_cards: List[str] = Field(default_factory=list)


class ColorOption(BaseModel):
    colors: List[str]
    name: str
    description: str
    strengths: List[str] = Field(default_factory=list)
    weaknesses: List[str] = Field(default_factory=list)


class CardRecommendation(BaseModel):
    card_name: str
    card_id: Optional[str] = None
    quantity: int = 1
    role: str
    reasoning: str
    image_uri: Optional[str] = None
    mana_cost: Optional[str] = None
    type_line: Optional[str] = None


class CardSlotGroup(BaseModel):
    slot_name: str
    description: str
    target_count: int
    recommendations: List[CardRecommendation] = Field(default_factory=list)


class LandRecommendation(BaseModel):
    card_name: str
    card_id: Optional[str] = None
    quantity: int
    category: str  # "dual", "utility", "basic", "fetch", "other"
    reasoning: str
    image_uri: Optional[str] = None


class SideboardRecommendation(BaseModel):
    card_name: str
    card_id: Optional[str] = None
    quantity: int
    target_matchups: List[str]
    reasoning: str
    image_uri: Optional[str] = None


# --- Request schemas ---

class StartGuidedBuildRequest(BaseModel):
    format: str = "standard"


class AdvanceStepRequest(BaseModel):
    session_id: UUID
    selections: Dict[str, Any] = Field(default_factory=dict)


# --- Response schemas ---

class StrategyStepData(BaseModel):
    archetypes: List[ArchetypeOption]
    meta_summary: str
    format: str


class ColorStepData(BaseModel):
    options: List[ColorOption]
    recommendation: str
    archetype: str


class CoreStepData(BaseModel):
    slots: List[CardSlotGroup]
    strategy_note: str
    cards_needed: int


class SupportStepData(BaseModel):
    slots: List[CardSlotGroup]
    current_deck_size: int
    remaining_nonland_slots: int


class ManaBaseStepData(BaseModel):
    lands: List[LandRecommendation]
    total_lands: int
    color_requirements: Dict[str, int]
    mana_curve_note: str


class SideboardStepData(BaseModel):
    recommendations: List[SideboardRecommendation]
    meta_matchups: List[Dict[str, Any]]
    sideboard_strategy: str


class ReviewStepData(BaseModel):
    deck_name: str
    archetype: str
    colors: List[str]
    strategy_summary: str
    main_deck: List[Dict[str, Any]]
    sideboard: List[Dict[str, Any]]
    main_deck_count: int
    sideboard_count: int
    validation_errors: List[str] = Field(default_factory=list)
    strengths: List[str] = Field(default_factory=list)
    weaknesses: List[str] = Field(default_factory=list)


class GuidedBuildStepResponse(BaseModel):
    session_id: UUID
    current_step: GuidedBuildStep
    step_index: int
    total_steps: int
    step_title: str
    step_description: str
    data: Dict[str, Any]
    ai_message: str


class GuidedBuildCompleteResponse(BaseModel):
    session_id: UUID
    deck_id: Optional[UUID] = None
    deck_name: str
    main_deck: List[Dict[str, Any]]
    sideboard: List[Dict[str, Any]]
    strategy_summary: str
    archetype: str
    colors: List[str]
    format: str
    is_valid: bool
    validation_errors: List[str] = Field(default_factory=list)
